core/lib/dal/src/tee_proof_generation_dal.rs

```diff
@@ -41,44 +38,72 @@ impl TeeProofGenerationDal<'_, '_> {
         let min_batch_number = min_batch_number.map_or(0, |num| i64::from(num.0));
         let query = sqlx::query!(
             r#"
-            UPDATE tee_proof_generation_details
-            SET
-                status = $1,
-                updated_at = NOW(),
-                prover_taken_at = NOW()
-            WHERE
-                tee_type = $2
-                AND l1_batch_number = (
-                    SELECT
-                        proofs.l1_batch_number
-                    FROM
-                        tee_proof_generation_details AS proofs
-                    JOIN
-                        tee_verifier_input_producer_jobs AS inputs
-                        ON proofs.l1_batch_number = inputs.l1_batch_number
-                    WHERE
-                        inputs.status = $3
-                        AND (
-                            proofs.status = $4
-                            OR (
-                                proofs.status = $1
-                                AND proofs.prover_taken_at < NOW() - $5::INTERVAL
-                            )
+            WITH
+            upsert AS (
+                SELECT
+                    proof_generation_details.l1_batch_number
+                FROM
+                    proof_generation_details
+                LEFT JOIN
+                    l1_batches
+                    ON proof_generation_details.l1_batch_number = l1_batches.number
+                LEFT JOIN
+                    tee_proof_generation_details AS proofs
+                    ON proof_generation_details.l1_batch_number = proofs.l1_batch_number
+                WHERE
+                    (
+                        proof_generation_details.vm_run_data_blob_url IS NOT NULL
+                        AND proof_generation_details.proof_gen_data_blob_url IS NOT NULL
+                        AND l1_batches.hash IS NOT NULL
+                        AND l1_batches.aux_data_hash IS NOT NULL
+                        AND l1_batches.meta_parameters_hash IS NOT NULL
+                    )
+                    AND (
+                        proofs.tee_type IS NULL
+                        OR proofs.tee_type = $1
+                    )
+                    AND (
+                        proofs.status IS NULL
+                        OR proofs.status = $3
+                        OR (
+                            proofs.status = $2
+                            AND proofs.prover_taken_at < NOW() - $4::INTERVAL
                         )
-                        AND proofs.l1_batch_number >= $6
-                    ORDER BY
-                        l1_batch_number ASC
-                    LIMIT
-                        1
-                    FOR UPDATE
-                    SKIP LOCKED
-                )
+                    )
+                    AND (
+                        proofs.l1_batch_number IS NULL
+                        OR proofs.l1_batch_number >= $5
+                    )
```

Comment:
This condition feels more natural for me (I hope it's correct):
```
                    AND (
                        (
                            proofs.tee_type IS NULL
                            AND proofs.status IS NULL
                            AND proofs.l1_batch_number IS NULL
                        ) OR (
                            proofs.tee_type = $1
                            AND (
                                proofs.status = $3
                                OR (
                                    proofs.status = $2
                                    AND proofs.prover_taken_at < NOW() - $4::INTERVAL
                                )
                            )
                            AND proofs.l1_batch_number >= $5
                        )
                    )
```
I think your variant evaluates to `true` in more cases, e.g. when `proofs.tee_type IS NULL AND proofs.status = $3 AND proofs.l1_batch_number >= $5`. It's probably not possible because of how DB schema is defined (`NOT NULL` columns), but it's not so obvious. 🤔

Reply:
```
                    AND (
                        (
                            proofs.tee_type IS NULL
                            AND proofs.status IS NULL
                            AND proofs.l1_batch_number IS NULL
                        ) OR (
                            proofs.l1_batch_number >= $5
                            AND (
                                proofs.tee_type != $1
                                OR
                                (
                                    proofs.tee_type = $1
                                    AND (
                                        proofs.status = $3
                                        OR (
                                            proofs.status = $2
                                            AND proofs.prover_taken_at
                                            < NOW() - $4::INTERVAL
                                        )
                                    )
                                )
                            )
                        )
                    )

```

Reply:
Pushed a changing commit with your version... I added pre filtering of the tee proof table with `tee_type` though, to exclude other TEE techs from the query.

Reply:
I believe [this line](https://github.com/matter-labs/zksync-era/pull/3017/commits/fa06b18517993810ee4ab3513029002ee2d57a21#diff-46d00b49fb68d6da13c9ecdc5ab3dab6adf502ec216349721586e96553652444R76) is now redundant.

BTW, one of the linter tests [is failing](https://github.com/matter-labs/zksync-era/actions/runs/11211617460/job/31160930751?pr=3017).